GitCommands/AsyncLoader.cs

```diff
@@ -56,135 +30,155 @@ public static Task<T> DoAsync<T>(Func<T> doMe, Action<T> continueWith, Action<As
         {
             AsyncLoader loader = new AsyncLoader();
             loader.LoadingError += (sender, e) => onError(e);
-            return loader.Load(doMe, continueWith);
+            return loader.LoadAsync(doMe, continueWith);
         }
 
         public static Task<T> DoAsync<T>(Func<T> doMe, Action<T> continueWith)
         {
             AsyncLoader loader = new AsyncLoader();
-            return loader.Load(doMe, continueWith);
+            return loader.LoadAsync(doMe, continueWith);
         }
 
         public static Task DoAsync(Action doMe, Action continueWith)
         {
             AsyncLoader loader = new AsyncLoader();
-            return loader.Load(doMe, continueWith);
+            return loader.LoadAsync(doMe, continueWith);
         }
 
-        public Task Load(Action loadContent, Action onLoaded)
+        public Task LoadAsync(Action loadContent, Action onLoaded)
         {
-            return Load((token) => loadContent(), onLoaded);
+            return LoadAsync((token) => loadContent(), onLoaded);
         }
 
-        public Task Load(Action<CancellationToken> loadContent, Action onLoaded)
+        public async Task LoadAsync(Action<CancellationToken> loadContent, Action onLoaded)
         {
             Cancel();
             _cancelledTokenSource?.Dispose();
             _cancelledTokenSource = new CancellationTokenSource();
             var token = _cancelledTokenSource.Token;
-            return Task.Factory.StartNew(() =>
+
+            try
+            {
+                if (Delay > TimeSpan.Zero)
                 {
-                    if (Delay > 0)
-                    {
-                        token.WaitHandle.WaitOne(TimeSpan.FromMilliseconds(Delay));
-                    }
+                    await Task.Delay(Delay, token).ConfigureAwait(false);
+                }
+                else
+                {
+                    await TaskScheduler.Default.SwitchTo(alwaysYield: true);
```

Comment:
@sharwell Could you please elaborate why you use here `SwitchTo(alwaysYield: true)`? Is there a general rule saying when it should be used?

Reply:
When you call an asynchronous method, the part of the method up to the first `await` always runs synchronously. To avoid this behavior, `Task.Run` can be used to force the operation to start on a different thread (or be queued). Since this would break threading rule 3, we instead switch to using `Run` or `RunAsync` to start asynchronous operations. However, `JoinableTaskFactory.RunAsync` behaves like other asynchronous methods, and doesn't yield prior to an explicit `await`. By calling `SwitchTo`, I do two things here:

1. I guarantee that the code afterwards is executing on the thread pool
2. If the code was *already* on the thread pool, I force the operation to yield and not continue synchronously

When switching from `Task.Run` to `JoinableTaskFactory.Run` (or `RunAsync`), the `SwitchTo` call is required to preserve semantics. If you don't need to force the code to execute as a separate operation, `SwitchTo` without the extra argument is more efficient.

Reply:
@sharwell's answer looks all correct except for one small detail. Using `Task.Run` is not a violation of the 3rd rule. But you can't synchronously block on it later (per the 3rd rule), so if you need to, you should have a JTF.RunAsync at this point, or in one of its callers and then you can await/Join on that later.